### PR TITLE
fix: ONNX conversion for Spark OneHotEncoder model

### PR DESCRIPTION
**Pull request by memoryz**
Fixing issue #551.

### CODE REVIEW

File: tests/sparkml/test_onehot_encoder.py [hunk 1]
```diff
@@ -19,30 +19,76 @@ class TestSparkmlOneHotEncoder(SparkMlTestCase):
...
+    def test_model_onehot_encoder_1(self):
+        """
+        Testing ONNX conversion for Spark OneHotEncoder when handleInvalid is set to "error" and dropLast set to False.
+        """
+        encoder = OneHotEncoder(inputCols=['index1', 'index2'], outputCols=['index1Vec', 'index2Vec'], handleInvalid="error", dropLast=False)
+        data = self.spark.createDataFrame([(0.0,5.0,), (1.0,4.0,), (2.0,3.0,), (2.0,2.0,), (0.0,1.0,), (2.0,0.0,)], ['index1','index2'])
         model = encoder.fit(data)
         model_onnx = convert_sparkml(
-            model, 'Sparkml OneHotEncoder', [('index', FloatTensorType([None, 1]))],
+            model, 'Sparkml OneHotEncoder', [('index1', FloatTensorType([None, 1])), ('index2', FloatTensorType([-1, 1]))],
```
Comment (xadupre) [inline tests/sparkml/test_onehot_encoder.py:30]: -1, 1 or None, 1 ?

File: requirements-dev.txt [hunk 2]
```diff
@@ -19,4 +19,4 @@ pytest-spark
 scikit-learn
 scipy
 wheel
-xgboost
+xgboost==1.5.2
```
[the remark this answers is not in the document]
Comment (memoryz) [reply requirements-dev.txt:22]: The latest xgboost version breaks some unit tests in the xgboost unit tests, which has nothing to do with my changes.

File: onnxmltools/convert/sparkml/operator_converters/onehot_encoder.py [hunk 3]
```diff
@@ -3,74 +3,76 @@
...
+            outputs=[encoded_feature_name],
+            op_domain="ai.onnx.ml",
+            op_version=1,
+            **encoder_attrs,
+        )
+
+        shape_variable_name = scope.get_unique_variable_name("shape_at_" + str(i))
+        container.add_initializer(shape_variable_name, onnx_proto.TensorProto.INT64, [2], [N, len(categories[i])])
+
+        container.add_node(
```
[the remark this answers is not in the document]
Comment (memoryz) [reply onnxmltools/convert/sparkml/operator_converters/onehot_encoder.py:59]: This is to match the behavior of the Spark OneHotEncoderModel. In Spark, if the OHEModel has multiple outputCols (one per category), the model does not concat the outputs into a vector. Instead you use a vector assembler to concat them, and the vector assembler will be translated into a concat ONNX operator.

File: tests/sparkml/test_onehot_encoder.py (hunk 1 above)
Comment (memoryz) [reply tests/sparkml/test_onehot_encoder.py:30]: Should be [None, 1]. I'll update the test code.

Comment (memoryz) [reply tests/sparkml/test_onehot_encoder.py]: Fixed.

**Review by xadupre — APPROVED** [review]

**Comment by memoryz** [comment]
@xadupre can you merge the change?

**Pull request merged**